### PR TITLE
Add ability to collect direct reclaim latency trace with ftrace

### PR DESCRIPTION
**Pull request by ethercflow**
It's now possible to collect linux kernel direct reclaim latency trace of all processes's alloc request.

 + Add ability to parse ftrace options
 + Add ability to stop subprocess with timeout both in python2 and python3
 + Add ability to trace direct reclaim latency trace

### CODE REVIEW

**Comment by ethercflow** [comment]
@AstroProfundis  @LinuxGit PTAL

File: insight.py [hunk 1]
```diff
@@ -31,18 +31,19 @@
 from measurement.files import logfiles
 from measurement.process import meta as proc_meta
 from measurement.tidb import pdctl
-
+from measurement.ftrace import ftrace
 
 class Insight():
     # data output dir
-    outdir = "data"
+    outdir = "%s/data" % os.getcwd()
```
Comment (AstroProfundis) [inline insight.py]: It's not necessary to `getcwd()` here, as in `fileutils.create_dir()` we use `os.makedirs()` to create/check target directory, which uses the (main) process's cwd for relative paths by default.

File: insight.py [hunk 2]
```diff
@@ -126,6 +127,22 @@ def run_perf(self, args):
...
+        if not args.ftrace:
+            logging.debug("Ignoring collecting of ftrace data.")
+            return
+        # perf requires root priviledge
+        if not util.is_root_privilege():
+            logging.fatal("It's required to run ftrace with root priviledge.")
+            return
+
+        if args.ftracepoint:
+            self.insight_ftrace = ftrace.InsightFtrace(vars(args))
```
Comment (AstroProfundis) [inline insight.py]: Would it be better to `var()` in `InsightFtrace`'s `__init__()`?

File: measurement/ftrace/ftrace.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    data_dir = "ftracedata"
+
+    def __init__(self, options={}):
+        self.ftrace_options = options
+
+    def run(self, outputdir=None):
+        ftrace_outputdir = fileutils.build_full_output_dir(
+            basedir=outputdir, subdir=self.data_dir)
+
+        if ftrace_outputdir is None:
```
Comment (AstroProfundis) [inline measurement/ftrace/ftrace.py]: `if not ftrace_outputdir`

File: measurement/ftrace/mem/drtracer.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        self.ftrace_options = options
+
+    def save_trace(self, outputdir=None):
+        _, stderr = util.run_cmd(["cd", self.tracefs])
+        if stderr:
+            logging.fatal("""ERROR: accessing tracing. Root user? Kernel has FTRACE?
+            debugfs mounted? (mount -t debugfs debugfs /sys/kernel/debug)""")
+            return
+
+        orginalwd = os.getcwd()
```
Comment (AstroProfundis) [inline measurement/ftrace/mem/drtracer.py]: We have a `util.cwd()` that may be useful. Or should we just put `originalwd` to a global variable (like in class `Insight`)? Will `chdir()` like this be usual in the future?

File: measurement/ftrace/mem/drtracer.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            self.ftrace_options and self.ftrace_options["ftrace_time"] else 60
+        full_outputdir = fileutils.build_full_output_dir(
+            basedir=outputdir, subdir=self.data_dir)
+        try:
+            _, stderr = util.run_cmd(["cat", "trace_pipe", ">", full_outputdir], timeout=time)
+            if stderr:
+                logging.fatal("ERROR: redirect trace_pipe failed")
+                os.chdir(originalwd)
+                return
+        except:
```
Comment (AstroProfundis) [inline measurement/ftrace/mem/drtracer.py]: What kind of exceptions do we ignore here? Is this mean that no matter how did the command fail, there's no need to inform user?

File: measurement/ftrace/mem/drtracer.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            _, stderr = util.run_cmd(["echo 1 > %s" % event], shell=True)
+            if stderr:
+                logging.fatal("ERROR: enable %s tracepoint failed" % event)
+                os.chdir(originalwd)
+                return
+
+        # collect trace
+        time = self.ftrace_options["ftrace_time"] if "ftrace_time" in \
+            self.ftrace_options and self.ftrace_options["ftrace_time"] else 60
+        full_outputdir = fileutils.build_full_output_dir(
```
Comment (AstroProfundis) [inline measurement/ftrace/mem/drtracer.py]: I think it's better to prepare output directory before actually doing trace, as if this step fails, there's no need to collect anything.

File: measurement/util.py [hunk 7]
```diff
@@ -146,3 +164,7 @@ def get_hostname():
     # This function is merely used, so only import socket package when necessary
     import socket
     return socket.gethostname()
+
+def python_version():
+    # get a numeric Python version
+    return sys.version_info[0] + sys.version_info[1] / 10
```
Comment (LinuxGit) [inline measurement/util.py]: major and minor are integers.
>>> sys.version_info
sys.version_info(major=2, minor=7, micro=10, releaselevel='final', serial=0)
>>> sys.version_info[0] + sys.version_info[1] / 10
2

File: measurement/util.py [hunk 8]
```diff
@@ -114,6 +123,15 @@ def parse_insight_opts():
     parser.add_argument("-v", "--verbose", action="store_true", default=False,
                         help="Print verbose output.")
 
+    parser.add_argument("-f", "--ftrace", action="store_true", default=False,
+                        help="Collect trace info using ftrace. Disabled by default.")
+    parser.add_argument("--ftracepoint",  action="store", default=None,
+                        help="Tracepoint to be traced")
```
Comment (LinuxGit) [inline measurement/util.py]: add period(.) in the end of `ftracepoint ` help message.
Which tracepoint does it currently support?

File: insight.py (hunk 1 above)
Comment (ethercflow) [reply insight.py]: If we call `fileutils.create_dir()` after `chdir()` with `outputdir=None`, the relative paths will not be the (main) process's cwd.

File: insight.py (hunk 2 above)
Comment (ethercflow) [reply insight.py]: I agree with you.

File: measurement/ftrace/mem/drtracer.py (hunk 5 above)
Comment (ethercflow) [reply measurement/ftrace/mem/drtracer.py]: In our case the except is `exception subprocess.TimeoutExpired` with python 3.5 and above.

File: measurement/util.py (hunk 7 above)
Comment (ethercflow) [reply measurement/util.py]: OK, I'll change to `sys.version_info[0] + sys.version_info[1] * 0.1`

File: measurement/util.py [hunk 9]
```diff
@@ -30,9 +31,17 @@ def cwd():
...
+    if python_version() >= 3.5:
+        p = Popen(cmd, shell=shell, timeout=timeout, stdout=PIPE, stderr=PIPE)
+        return p.communicate()
+    elif not timeout:
+        p = Popen(cmd, shell=shell, stdout=PIPE, stderr=PIPE)
+        return p.communicate()
+
+    p = Popen(cmd, shell=shell)
+    time.sleep(timeout)
+    return _, _
```
Comment (LinuxGit) [inline measurement/util.py]: I don't find timeout in Popen Constructor: https://docs.python.org/3.5/library/subprocess.html#subprocess.Popen

what if python < 3.5 and timeout is set? why sleep timeout

File: insight.py [hunk 10]
```diff
@@ -51,7 +53,7 @@ def __init__(self, args):
         if not args.alias:
             self.alias = util.get_hostname()
         self.full_outdir = fileutils.create_dir(
-            os.path.join(self.outdir, self.alias))
+            os.path.join(self.cwd, self.outdir, self.alias))
```
Comment (AstroProfundis) [inline insight.py]: What if a absolute path is passed in with `--output`? Maybe we can check with `str.startswith('/')`?

File: measurement/ftrace/mem/drtracer.py [hunk 11]
```diff
@@ -23,59 +23,49 @@ class DirectReclaimTracer():
     def __init__(self, options={}):
         self.ftrace_options = options
 
-    def save_trace(self, outputdir=None):
+    def save_trace(self, cwd, outputdir):
```
Comment (AstroProfundis) [inline measurement/ftrace/mem/drtracer.py]: I think it's better to check for empty values anyway.

**Pull request merged**